### DISCUSSION TITLE
How to Add favicon.ico on Next.js 13 & Next.js 14

### DISCUSSION THREAD

**Discussion by clevermiraz** [vercel/next.js#50704] [Help] [answered]
### Summary

I wanted to Add favicon.ico on my website how can i do that

**Comment by devjiwonchoi** [comment, marked as answer]
# Edit

Everyone is getting confused, let me make this simple.

## Convention

Place `favicon.ico` inside the `app/` or `public/` folder, which Next.js will auto-generate the necessary metadata.

```
.
├── app/
│   └── favicon.ico <-- here OR
└── public/
    └── favicon.ico <-- here
```

## Else

Place the favicon file(s) on the public, then use `Metadata` or `generateMedatada`.

```js
export const metadata = {
  icons: {
    icon: '/icon.png', // /public path
  },
}
```

You can also change favicon by color scheme:

```ts
icons: {
  icon: [
    {
      url: '/light-icon.png',
      media: '(prefers-color-scheme: light)',
    },
    {
      url: '/dark-icon.png',
      media: '(prefers-color-scheme: dark)',
    },
  ],
},
```

### Reference

- [generateMetadata#icons](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#icons)

**Reply by CMCDragonkai** [reply]
This no longer works. Putting it into the `public/` directory doesn't result in automatic creation of the `<link>` elements in the header. It only works if you put it into `src/app/` directory.

**Reply by CMCDragonkai** [reply]
And the nextjs docs makes no mention of `public/` as a location for favicon, or icon or apple-touch-icon.

**Reply by devjiwonchoi** [reply]
@CMCDragonkai Right, I'll update the answer. Thank you!

**Reply by CMCDragonkai** [reply]
But there's no answer for how to put the images into the `public/` directory. You now have to manually setup the `<link>` elements by exporting `metadata: Metadata` with the `icons` property.

**Reply by devjiwonchoi** [reply]
> Next.js will evaluate the file and automatically add the appropriate tags to your app's <head> element.

If you mean favicon image, Next.js will do the job for you if inside `/app`, not getting why you'd want to go for `/public`.

**Reply by CMCDragonkai** [reply]
Cause I got a lot of icon sizes and it's better to put it in public where all my other files are. Also why so much backwards incompatible changes? Keep it simple and conventional.

**Reply by devjiwonchoi** [reply]
To make it clear, it is not breaking any conventions. It's your choice to put on public vs app, it's just that in the `/app` it kindly automates metadata.

**Reply by lucgagan** [reply]
What if I want to use png as an icon?

**Reply by ralphestor** [reply]
i just renamed the png file as 'favicon.ico' inside the app directory. it worked for me.

**Reply by CalvinFrank** [reply]
I tried this and it doesn't work for me.

**Reply by HectorOsborneRodriguez** [reply]
Just tried this solution with "next": "14.0.4" and it didn't work. 
It's either favicon.ico or nothing.

**Reply by pablohcs7** [reply]
@HectorOsborneRodriguez Try renaming favicon.ico to icon.ico using next 14

**Reply by ruck-45** [reply]
renaming favicon.icon as icon.ico in public folder and including 
icons: {
    icon: '/icon.png', // /public path
  },
  
 in metadata in layout.tsx worked for me

**Reply by furqanmirzaa** [reply]
Renamed the file to icon.png and put the file in the root of public and then add
icons: {
    icon: "/icon.png",
  },
in the metadata of layout.js worked for me.

**Reply by imran-khani** [reply]
Thank you. it worked for me.

**Reply by guisilva19** [reply]
It worked, Thanks!

**Reply by Apongpoh** [reply]
Great! this work for me

**Reply by devDanielCespedes** [reply]
I love you ❤️

**Reply by jonbrunt** [reply]
Still working in 2024 with Next 14. Thanks @eclipssed !

**Reply by Marius9595** [reply]
Thanks, I used the convention one.

 **Important!** Your file has to be named as favicon.ico otherwise :D

**Reply by StackOverflowIsBetterThanAnyAI** [reply]
Thanks, I used the method you labelled 'else'. Don't forget to use the correct name of your icon (favicon.ico).

**Reply by charlie763** [reply]
I have put my favicon in the root app directory, and it keeps getting overwritten with the default next favicon whenever I run `npm run build`. Perhaps because I have a custom static export config? -
```
const nextConfig = {
  assetPrefix: '/lp', // This prepends /lp to any hrefs in the static content
  basePath: '/lp', // This prepends /lp to the path for local development 
  distDir: 'out/lp', // This makes sure static content is exported to /out/lp
  output: 'export', // DO NOT CHANGE THIS - this ensures the site is static
  sassOptions: {
    includePaths: [path.join(__dirname, 'styles')],
  },
};
```

any thoughts @devjiwonchoi ?

**Reply by atlamors** [reply]
Hey @charlie763 I just ran into the same issue.

I was able to resolve it by moving the favicon into the /public/ dir and deleting all other copies of it.

Hope this helps you.

**Reply by victornasar** [reply]
@atlamors same here. I was struggling with this for an hour! I didn't even think about moving it to the public folder.

**Reply by clevermiraz** [reply]
We can also use this in the main `layout.tsx`.

assume that we have logo in public directory

```js
export const metadata = {
    title: "Khana Khazana",
    description: "The best food in town",
    icons: {
        icon: [
            {
                url: "/logo.svg", // /public path
                href: "/logo.svg", // /public path
            },
        ],
    },
};
```

**Comment by clevermiraz** [comment]
Yup, Just Put `favicon.ico` on the root of the app directory. Then do a hard refresh. Next js will automatically detect what to do.

thanks @devjiwonchoi

**Reply by ChristophHandschuh** [reply]
keep in mind that the favicon is cached. When you dont see it make an request to the url of the favicon and it should appear

**Reply by amitsuthar69** [reply]
You saved me hours man! Just a hard refresh was all I needed.

**Reply by viktoriabakun** [reply]
@ChristophHandschuh thank you!

**GUYS, HARD RELOAD!** 

<img width="354" alt="1024KB" src="https://github.com/vercel/next.js/assets/59821235/86f5a2ae-4fe7-4f8e-8ac6-ae2ea92ef153">

**Reply by MdHasib01** [reply]
@viktoriabakun thanks a lot!!!

**Comment by lumenwrites** [comment]
I want to put my `favicon.ico` into a different directory. How can I customize that?

**Reply by clevermiraz** [reply]
then put favicon.ico in public directory. Then Try a shit.

and you can access by 

```js
<link rel="icon" type="image/x-icon" href="./favicon.ico">
```

**Comment by wadakatu** [comment]
@lumenwrites @CMCDragonkai 

Hi.
I managed to create link tag of favicon with Metadata API.
I believe you can do it even though your favicon file is in the public folder.

The successful code is something like below.

```js
import Favicon from '/public/images/Metadata/favicon.ico';

export const metadata: Metadata = {
  title: 'Next.js',
  description: 'Generated by create next app',
  icons: [{ rel: 'icon', url: Favicon.src }],
};
```

Hope it helps.

**Reply by CMCDragonkai** [reply]
Interesting, I just ended up loading from `/favicon.ico` directly and it resolved from `public` anyway.

**Reply by Kenny4297** [reply]
I tried EVERYTHING and this was the only solution that worked! Thanks so much!

**Reply by KCTWLSK** [reply]
really interesting that this fantastic solution would work whilst the official ones would not...

**Reply by mnsuccess** [reply]
Thanks !! This was the only solution that worked for me within Nextjs 14

**Reply by Rinya7** [reply]
Thanks. Its work. Nextjs 14

**Reply by kir4che** [reply]
thx. It's work!!!

**Reply by ileys-admin** [reply]
Only thing that worked, this is way to unneccesarily difficult, the entire point of why it wasnt showing up was because i was creating a mutilingual site whose root is at [lang] so it never really ever finds the favicon at the root cuz its always redirected. and I followed that example directly from the nextjs website 

also I just realized how stupid I was and anyone who is going through my same issue do the above or modify the config matchter inside the middleware to exclude the favicon.ico which nextjs automatically detects
`export const config = {
  matcher: [
    // Skip all internal paths (_next)
    "/((?!api|_next/static|_next/image|favicon.ico).*)",
    // Optional: only run on root (/) URL
    // '/'
  ],
};`

**Reply by narutophuc113** [reply]
Only thing that worked for me.

I set up the internationalize and configed the matcher in middleware, but only this way worked.

**Reply by CarlosSousa2001** [reply]
Thanks !!

**Reply by Wellington-Matoss** [reply]
Thanks, man!!

**Reply by Vdcds** [reply]
Thanks, worked.

**Reply by Osvelsx** [reply]
This one worked for me.

Thanks friend, I was facing the same error but I just deployed to PD and now the icon is showing.

**Reply by krisgo1** [reply]
Thanks - as a person moving from /pages to /app, I have no idea why such a simple thing has become so complicated...

**Comment by Basovs** [comment]
Sometimes after putting the favicon.ico into /app or /public + "link" element you still need to reset the browser cache for the icon to be displayed. Try reset brower with cmd+q on at least, dont know how it works on Windows. Thats what helped me.

**Reply by ChrisHPZ** [reply]
On Chrome, we can hard reload by pressing the Ctrl and F5 keys together. This applies to Windows PC's. We can also open Chrome's console with the F12 key, then hold down on the Reload icon and click on "Clear Cache and Hard Reload" option

**Reply by ddkeating** [reply]
Thanks this worked for me!

**Reply by ShahAashka** [reply]
Worked for me as well. thanks @ChrisHPZ

**Comment by ChrisHPZ** [comment]
I just added an /app directory in my project root that contains a favicon.ico file. It just worked out of the box. I didn't need to import or set up any additional components.

**Reply by didiamuri** [reply]
Here is the solution that has a market for me :

```
This is the simplest way to use Favicon in the Next.js 13.4 Stable version.

1. You need to convert your jpg, png, or SVG to .ico. It takes only a few seconds.
2. Rename your .ico file to icon.ico and paste it into the app folder and remove the favicon.ico from the app folder
```

source: [https://stackoverflow.com/a/76412379/9487515](https://stackoverflow.com/a/76412379/9487515)

**Reply by jeevanlogiciel** [reply]
But ico files are larger in size as compared to png

**Comment by getFrontend** [comment]
In Next 14, just rename the **favicon** to **icon** and put it in the **app** folder.

Everything will automatically work the way it should! You don't need to look for other solutions anymore)

**Reply by Cnerd-Mahadi** [reply]
thanks man! exact answer I was looking for <3

**Reply by NicollasMelo** [reply]
Thanks man!

**Reply by ParvinEyvazov** [reply]
Thank you!

**Reply by Choi-Jiwon-38** [reply]
@devjiwonchoi
 In Next.js version 14, when customizing the favicon under the app directory as `favicon.ico`, I encountered an issue where the transparent background of the `.ico` file was not applied.
 Additionally, in the production environment via Vercel, the favicon was not applied at all. However, when I renamed the file to `icon.ico`, both the issue with the transparent background and the problem in the production environment were resolved.
  The official documentation seems to recommend using `favicon.ico`, but why might these issues occur?

**Reply by devjiwonchoi** [reply]
Hey @Choi-Jiwon-38 Jiwon! 🙌
Could you open an issue with a reproduction?
Feel free to ping me on it and I'd happy to take a look. 😄

**Comment by reinaldo-teixeira** [comment]
I was struggling with this problem on MacOs and this help me https://stackoverflow.com/a/76301605

**Comment by agenciakls-web** [comment]
With me the file don't working more with favicon in Vercel.

I using now `app/` with name `icon.ico` 

Reference: [This Links](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/app-icons)

Using `favicon.ico` working in my first server, in news servers of clients my don't working!

I'm Brazilian!

**Comment by paulenokels** [comment]
If your favicon is not in `ico` format. You can do this instead:    ` <link rel="icon" type="image/x-icon" href="./favicon.svg" />`

**Comment by Bamof25th** [comment]
### Fixes which worked for me 
- Just put favicon.iso in your root directory, and that works . 
- Else just rename the favicon.ico on  /app folder to icon.ico  .

**Comment by didiamuri** [comment]
You can even use an svg icon by specifying it in the main layout in this way:

"next": "14.2.7",

![metadata](https://github.com/user-attachments/assets/fb7d6a44-4274-4315-b6fd-0e0bf88cced2)


And everything works fine.

**Comment by magrega** [comment]
Guys, how can I add an svg icon as a favicon? As per doc, I tried moving it into /app and /public directory with icon name of the file but it doesnt work. In network browser tab i have a 404 error for favicon.ico

in _document.tsx i have this:

```
import { Html, Main, NextScript } from 'next/document'
import Head from 'next/head'

export const metadata = {
	icons: {
		icon: '/favicon.svg'
	}
}

export default function Document() {
	return (
		<Html lang='ru'>
			<Head>
				<link rel='icon' href='/favicon.svg' />
				<meta name='apple-mobile-web-app-status-bar-style' content='#ff7652' />
			</Head>
			<body>
				<Main />
				<NextScript />
			</body>
		</Html>
	)
}

```

**Reply by didiamuri** [reply]
> You can even use an svg icon by specifying it in the main layout in this way:
> 
> "next": "14.2.7",
> 
> ![metadata](https://private-user-images.githubusercontent.com/31784455/366545697-fb7d6a44-4274-4315-b6fd-0e0bf88cced2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNjQ4MTksIm5iZiI6MTcyOTE2NDUxOSwicGF0aCI6Ii8zMTc4NDQ1NS8zNjY1NDU2OTctZmI3ZDZhNDQtNDI3NC00MzE1LWI2ZmQtMGUwYmY4OGNjZWQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE3VDExMjgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmZjE4MTkwOGYzYmJmZmU1NTgyZTA5NDBlNmM1ZWE1OWIwNjYyNmZkZDQ3NzY5ODg4ZDY3NzIzZDg4ZGE1NWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HK50GiXLQKZ4wTkX3RP49AD7Mj4n1gkXmVc4nEHAQps)
> 
> And everything works fine.

**Comment by paras-verma7454** [comment]
Just delete the favicon.ico and paste your favicon in the `app` folder and rename your new favicon to favicon.ico 
and it will work
![image](https://github.com/user-attachments/assets/9dfe9aa1-b61a-4f53-8f38-b2fed4f8c400)

**Comment by odilson-dev** [comment]
This was really helpful. Thank you❤️

**Comment by phbernard** [comment]
You can use a dedicated tool to [generate a favicon for Next.js](https://realfavicongenerator.net/favicon-generator/nextjs). It lets you create and design icons platform per platform. For the setup, all you have to do is dropping files in the proper directories:

![image](https://github.com/user-attachments/assets/00c2a901-f07e-4ba6-be30-202161762c72)

It follows the [official conventions](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/app-icons), so you really setup your favicon the Next.js way, as described by @devjiwonchoi .

Full disclosure: I'm the author of this service. I realize posting this here is borderline, but I really think it can genuinely help.

**Reply by lukylab** [reply]
good job!
